release/train_tests/benchmark/recsys/criteo.py

```diff
@@ -0,0 +1,340 @@
+        "valid": "s3://ray-benchmark-data-internal/criteo/tsv.gz/valid/",
+        "test": "s3://ray-benchmark-data-internal/criteo/tsv.gz/test/",
+    }
+    ds_path = stage_to_path[stage]
+
+    ds = ray.data.read_csv(
+        ds_path,
+        read_options=pyarrow.csv.ReadOptions(column_names=DEFAULT_COLUMN_NAMES),
+        parse_options=pyarrow.csv.ParseOptions(delimiter="\t"),
+        shuffle=("files" if stage == "train" else None),  # coarse file-level shuffle
+    )
+
+    # Convert categorical features to integers.
+
+    # Option 1: Use the built-in preprocessor (This is not performant enough.)
+    # from ray.data.preprocessors import OrdinalEncoder
+    # categorical_to_indices_preprocessor = OrdinalEncoder(columns=DEFAULT_CAT_NAMES)
+    # ds = categorical_to_indices_preprocessor.fit_transform(ds)
+
+    # Option 2: groupby() w/ hash-based shuffle
+
+    # Fetch cached value counts instead of "fitting" the preprocessor from scratch.
+    COMPUTE_VALUE_COUNTS_FROM_SCRATCH: bool = False
+
+    FREQUENCY_THRESHOLD = 3
+    LOW_FREQUENCY_INDEX = 1  # map low frequency values -> 1
+    categorical_value_to_index = {}
+    for cat_feature in DEFAULT_CAT_NAMES:
+        if COMPUTE_VALUE_COUNTS_FROM_SCRATCH:
+            logger.info(f"Computing value counts for: {cat_feature}")
+
+            def filter_features(batch):
+                features = batch[cat_feature]
+                features[features == None] = ""
+                return {cat_feature: features}
+
+            value_counts = [
+                (group[cat_feature], group["count()"])
+                for group in ds.map_batches(filter_features)
+                .groupby(key=cat_feature)
+                .count()
+                .take_all()
+            ]
+        else:
+            json_filepath = CAT_FEATURE_VALUE_COUNT_JSON_PATH_PATTERN.format(
+                cat_feature
+            )
+            logger.info(f"Downloading value counts file: {json_filepath}")
+            value_counts = read_json_from_s3(bucket_name=S3_BUCKET, key=json_filepath)
+
+        value_counts = filter(lambda x: x[1] >= FREQUENCY_THRESHOLD, value_counts)
+        categorical_value_to_index[cat_feature] = {
+            val: i for i, (val, _) in enumerate(value_counts, start=2)
+        }
+
+    # This mapping is large, so put a shared copy in the object store for all the map tasks to use.
+    categorical_value_to_index_ref = ray.put(categorical_value_to_index)
+
+    # Clean data.
+    ds = ds.map_batches(fill_missing)
+
+    def categorical_values_to_indices(
+        batch: Dict[str, np.ndarray], mapping_ref: ray.ObjectRef
+    ):
+        mapping: Dict[str, int] = ray.get(mapping_ref)
+        for cat_feature in DEFAULT_CAT_NAMES:
+            batch[cat_feature] = np.vectorize(
+                lambda k: mapping.get(cat_feature, {}).get(k, LOW_FREQUENCY_INDEX)
+            )(batch[cat_feature])
+        return batch
+
+    ds = ds.map_batches(
```

Comment:
since we are already precomputing `categorical_values_to_index`.
Is it a good idea to also precompute `categorical_values_to_indices` and ` concat_and_normalize_dense_features`?

Reply:
I was thinking the `categorical_values_to_index` is the "preprocessor" that's being fit. The `concat_and_normalize_dense_features` is the online preprocessing that can be done so that we don't need to store the whole dataset again in this intermediate format.